### PR TITLE
No longer silently hide errors in Metal completion handlers (alternative approach)

### PR DESCRIPTION
**Pull request by shoaibkamil**
As discussed in the dev meeting, here is an alternative approach for fixing #7780.  Now, the runtime saves an error during the callback from a failed command buffer, and reports the error back at the next opportunity for reporting.  Note that this means that a pipeline may get an error that should be attributed to the previous pipeline, but additional error information can be added (e.g. prepending the error message with the name of the kernel that caused it) if wanted.

### CODE REVIEW

File: src/runtime/metal.cpp [hunk 1]
```diff
@@ -426,9 +426,26 @@ WEAK command_buffer_completed_handler_block_descriptor_1 command_buffer_complete
     0, sizeof(command_buffer_completed_handler_block_literal)};
 
 WEAK void command_buffer_completed_handler_invoke(command_buffer_completed_handler_block_literal *block, mtl_command_buffer *buffer) {
```
Comment (steven-johnson) [inline src/runtime/metal.cpp:428]: Is there no way to pass a valid user_context to this callback?

File: test/correctness/gpu_metal_completion_handler_error_check.cpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    f.gpu_tile(x, c, xi, ci, 4, 4);
+    f.update(0).gpu_tile(r.x, c, rxi, ci, 4, 4);
+
+    // Because the error handler is invoked from a Metal runtime thread, setting a custom handler for just
+    // this pipeline is insufficient.  Instead, we set a custom handler for the JIT runtime
+    JITHandlers handlers;
+    handlers.custom_error = my_error;
+    Internal::JITSharedRuntime::set_default_handlers(handlers);
+
+    f.realize({1000, 100}, t);
```
Comment (steven-johnson) [inline test/correctness/gpu_metal_completion_handler_error_check.cpp:42]: Will this also work if it is a `Callable` and the caller is relying on an error code being returned (vs halide_error being called)? If so, I don't understand how.

File: src/runtime/metal.cpp (hunk 1 above)
Comment (shoaibkamil) [reply src/runtime/metal.cpp:428]: I can enable that, sure.  But I'm not quite clear what should happen inside the callback then, with the `user_context`.  Is the idea users would override this with their own implementation and do stuff with the `user_context`?

File: test/correctness/gpu_metal_completion_handler_error_check.cpp (hunk 2 above)
Comment (shoaibkamil) [reply test/correctness/gpu_metal_completion_handler_error_check.cpp]: Sorry, I didn't update the branch before opening this alternative.

File: src/runtime/metal.cpp (hunk 1 above)
Comment (steven-johnson) [reply src/runtime/metal.cpp:480]: Yes. Code in google does this regularly... though the print and error handlers don't really use user_context, it's good form to pass in a valid value at all times unless there's a compelling reason not to.

Comment (shoaibkamil) [reply src/runtime/metal.cpp:480]: Ok, let me remember how to stash a `void*` pointer in the right place.  I think users would likely override with an Objective C implementation, but makes sense that we should pass it through.

**Comment by shoaibkamil** [comment]
Sorry @steven-johnson, I brain-farted and hadn't pushed the branch I intended to push when I opened this.  Should be correct now.

**Comment by slomp** [comment]
@shoaibkamil I recommend pre-pending the error message with something that reminds the user that this error has happened unexpectedly during the execution of a previous pipeline.

**Comment by steven-johnson** [comment]
`C:\build_bot\worker\halide-testbranch-main-llvm19-x86-64-windows-cmake\halide-source\test\correctness\gpu_metal_completion_handler_error_check.cpp(2): fatal error C1083: Cannot open include file: 'unistd.h': No such file or directory
`

**Comment by shoaibkamil** [comment]
@steven-johnson I've substantially revamped the API here to enable users to pass a user context.  PTAL

**Review by steven-johnson — APPROVED** [review]
LGTM pending green buildbots

**Comment by shoaibkamil** [comment]
Investigating crash, and will add tests to makefile.

**Comment by shoaibkamil** [comment]
All green finally, modulor vulkan errors.

**Pull request merged**